package/src/bin/thing-flinger.rs

```diff
@@ -0,0 +1,590 @@
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn copy_package_manifest_to_staging(
+    config: &Config,
+    builder: &Server,
+    destination: &Server,
+) -> Result<()> {
+    let mut path = PathBuf::from(&config.builder.omicron_path);
+    path.push("package-manifest.toml");
+    let cmd = format!(
+        "rsync {} {}@{}:~/{}",
+        path.to_string_lossy(),
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn run_omicron_package_from_staging(
+    config: &Config,
+    destination: &Server,
+    artifact_dir: &Path,
+    install_dir: &Path,
+) -> Result<()> {
+    let mut deployment_src = PathBuf::from(&config.deployment.staging_dir);
+    deployment_src.push(&artifact_dir);
+
+    // Run `omicron-package install` on the deployment server
+    let cmd = format!(
+        "cd ~/{} && pfexec ./omicron-package install --in ~/{} --out {}",
+        config.deployment.staging_dir,
+        deployment_src.to_string_lossy(),
+        install_dir.to_string_lossy()
+    );
+    println!("$ {}", cmd);
+    ssh_exec(destination, &cmd, true)
+}
+
+fn copy_overlay_files_to_staging(
+    config: &Config,
+    pkg_dir: &str,
+    builder: &Server,
+    destination: &Server,
+    destination_name: &str,
+) -> Result<()> {
+    let cmd = format!(
+        "rsync -avz {}/overlay/{}/ {}@{}:~/{}/overlay/",
+        pkg_dir,
+        destination_name,
+        destination.username,
+        destination.addr,
+        config.deployment.staging_dir
+    );
+    println!("$ {}", cmd);
+    ssh_exec(builder, &cmd, true)
+}
+
+fn install_overlay_files_from_staging(
+    config: &Config,
+    destination: &Server,
+    install_dir: &Path,
+) -> Result<()> {
+    let cmd = format!(
+        "pfexec cp -r ~/{}/overlay/* {}",
+        config.deployment.staging_dir,
+        install_dir.to_string_lossy()
+    );
+    println!("$ {}", cmd);
+    ssh_exec(&destination, &cmd, false)
+}
+
+// For now, we just restart sled-agent, as that's the only service with an
+// overlay file.
+fn restart_services(destination: &Server) -> Result<()> {
+    ssh_exec(destination, "svcadm restart sled-agent", false)
+}
+
+fn dir_string(dirs: &[PathBuf]) -> String {
+    dirs.iter().map(|dir| dir.to_string_lossy().to_string() + " ").collect()
+}
+
+fn dir_per_deploy_server(config: &Config, root: &Path) -> Vec<PathBuf> {
+    config
+        .deployment
+        .servers
+        .iter()
+        .map(|server_dir| {
+            let mut dir = PathBuf::from(root);
+            dir.push(server_dir);
+            dir
+        })
+        .collect()
+}
+
+fn ssh_exec(
+    server: &Server,
+    remote_cmd: &str,
+    forward_agent: bool,
+) -> Result<()> {
+    // Source .profile, so we have access to cargo. Rustup installs knowledge
+    // about the cargo path here.
+    let remote_cmd = String::from(". ~/.profile && ") + remote_cmd;
```

Comment:
Same here, does `~` need to be the same as `server.username`?

Reply:
Yes, it does. I was not super happy with having to do this, but it was unfortunately necessary. Again, I can change to `$HOME` if you feel strongly, although I have no indication that that is any more stable than `~`

Reply:
Changed to use $HOME.